[ONNX] Enable attribute type checking in onnx dispatcher

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100554
* __->__ #105104

The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

titaiwangms committed:
[ONNX] Enable attribute type checking in onnx dispatcher

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Enable attribute type checking in onnx dispatcher"


The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -494,11 +494,20 @@ def perfect_match_inputs(
                 return False
-        # TODO: attributes dtype is not supported right now, so we just check their existence.
-        if set(function_attributes) != self.attributes:
+        # Check attributes numbers are the same
```
```suggestion
        # Check attributes keys are the same
```

BowenBao replied:
or names

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -494,11 +494,20 @@ def perfect_match_inputs(
+                type(attribute)
+            )
+            if attribute_onnx_type != self.attributes[attribute_name].type:
+                # If the attribute type of the OpSchema and the attribute type don't match,
```
Will this reasoning be visible in diagnostic?

titaiwangms replied:
I updated the rule description to include attribute dtype

justinchuby replied:
When the input is int and the signature is float. Do we consider them compatible? I tend to think we should

BowenBao replied:
> I updated the rule description to include attribute dtype

I was thinking about accumulating very fine grained diagnostic message. Like adding `this is not perfect match because attribute {name} has {actual_type} vs {expected_type}.` into additional messages. Similarly for argument checks.

Doesn't need to be in this PR if these are not done before.

titaiwangms replied:
> When the input is int and the signature is float. Do we consider them compatible? I tend to think we should

Should we make this exception?

titaiwangms replied:
> > I updated the rule description to include attribute dtype
> 
> I was thinking about accumulating very fine grained diagnostic message. Like adding `this is not perfect match because attribute {name} has {actual_type} vs {expected_type}.` into additional messages. Similarly for argument checks.
> 
> Doesn't need to be in this PR if these are not done before.

https://github.com/pytorch/pytorch/issues/105174

justinchuby replied:
> > When the input is int and the signature is float. Do we consider them compatible? I tend to think we should
> 
> Should we make this exception?

Python / PyTorch treats int and float pretty loosely so I think it’s reasonable.  PyTorch’s Scalar type can be an int or float. We in the graph building process will convert int to float if the attribute type is float

justinchuby replied:
So I think it’s reasonable to be able to match. Otherwise we can also fix that in a pass?

titaiwangms replied:
If there is an overload relying on this attribute dtype (int or float), this would be an issue.

justinchuby replied:
That's true. I guess the fuzzy match logic can handle it for now then?

BowenBao commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -38,6 +43,23 @@ def from_torch_dtype_to_onnx_dtype_str(dtype: Union[torch.dtype, type]) -> Set[s
+) -> Optional[onnx.defs.OpSchema.AttrType]:
+    import onnx.defs  # type: ignore[import]
+
+    _PYTHON_TYPE_TO_ONNX_ATTRIBUTE_TYPE = {
```
bool?

titaiwangms replied:
Bool is actually not an AttributeProto type

https://github.com/onnx/onnx/blob/547032edb8f86adb1b38d37f0a87aa61ee2ff580/onnx/onnx.proto#L118-L139

justinchuby replied:
bool -> onnx.defs.OpSchema.AttrType.INT

titaiwangms replied:
I am surprised that we don't have any test cases using bool attribute.

justinchuby replied:
Dropout maybe?

titaiwangms replied:
I can try adding it as a test case here

titaiwangms replied:
Added a test in `test_fx_to_onnx.py` showing that although attribute type checking is added, the aten::convolution which has sequence/bool type in attribute can still be found as perfect match.

BowenBao commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -38,6 +43,23 @@ def from_torch_dtype_to_onnx_dtype_str(dtype: Union[torch.dtype, type]) -> Set[s
     return _TORCH_DTYPE_TO_COMPATIBLE_ONNX_TYPE_STRINGS[dtype]
 
 
+def from_python_type_to_onnx_attribute_type(
```
Question for my understanding.

- Where do we convert this python object to onnx attribute? I assume eventually we need to do this conversion.
- If my assumption above is correct, Can we merge the core of the two logics? Feels like much can be reused.
- If above is achievable, we should be able to extend support for `torch.dtype`, `torch.device`, etc as part of the python types supported here?

titaiwangms replied:
> Where do we convert this python object to onnx attribute? I assume eventually we need to do this conversion.

For most of attribute (int, float, bool, str), we don't need to convert them, we only need to `filter_incompatible_and_dtype_convert_kwargs` to omit useless kwargs and JitScalarType convert torch dtype to onnx dtype. The AttributeProto is written into OpSchema with the help of param_schema(), but it was never used.

justinchuby replied:
The conversion is handled by torchscript if I understand your question right

titaiwangms committed:
Update on "[ONNX] Enable attribute type checking in onnx dispatcher"


The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Enable attribute type checking in onnx dispatcher"


The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Enable attribute type checking in onnx dispatcher"


The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Enable attribute type checking in onnx dispatcher"


The dipatcher didn't check attribute dtype, as AttributeProto is a totally different system from InputProto in ONNX. This PR introduces the mapping table for AttributeProto type to python type. And further utilize it in opschema matching.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
